spring-rabbit-test/src/test/java/org/springframework/amqp/rabbit/test/ExampleListenerIntegrationTest.java

```diff
@@ -0,0 +1,159 @@
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.springframework.amqp.core.AnonymousQueue;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.annotation.EnableRabbit;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
+import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.messaging.handler.annotation.Payload;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+@ContextConfiguration
+@RunWith(SpringJUnit4ClassRunner.class)
+@DirtiesContext
+public class ExampleListenerIntegrationTest {
+
+	@Rule
+	public BrokerRunning brokerRunning = BrokerRunning.isRunning();
+
+	@Autowired
+	private RabbitTemplate rabbitTemplate;
+
+	@Autowired
+	private Queue queue1;
+
+	@Autowired
+	private Queue queue2;
+
+	@Autowired
+	private RabbitListenerTestHarness harness;
+
+	@Test
+	public void testTwoWay() throws Exception {
+		assertEquals("FOO", this.rabbitTemplate.convertSendAndReceive(this.queue1.getName(), "foo"));
+
+		BlockingQueue<Object[]> argumentQueue = this.harness.getArgumentQueueFor("foo");
+		assertNotNull(argumentQueue);
+		Object[] args = argumentQueue.poll(10, TimeUnit.SECONDS);
+		assertThat((String) args[0], equalTo("foo"));
+		BlockingQueue<Object> resultQueue = this.harness.getResultQueueFor("foo");
+		assertNotNull(resultQueue);
+		Object result = resultQueue.poll(10, TimeUnit.SECONDS);
+		assertThat((String) result, equalTo("FOO"));
+	}
+
+	@Test
+	public void testOneWay() throws Exception {
+		this.rabbitTemplate.convertAndSend(this.queue2.getName(), "bar");
+		this.rabbitTemplate.convertAndSend(this.queue2.getName(), "baz");
+
+		BlockingQueue<Object[]> argumentQueue = this.harness.getArgumentQueueFor("bar");
+		assertNotNull(argumentQueue);
+		Object[] args = argumentQueue.poll(10, TimeUnit.SECONDS);
+		assertThat((String) args[0], equalTo("bar"));
+		assertThat((String) args[1], equalTo(queue2.getName()));
+
+		args = argumentQueue.poll(10, TimeUnit.SECONDS);
+		assertThat((String) args[0], equalTo("baz"));
+		assertThat((String) args[1], equalTo(queue2.getName()));	}
+
+	@Configuration
+	@EnableRabbit
+	@RabbitListenerIntegrationTest
```

Comment:
>Maybe ...

Yes - that's my goal - I already started writing a `MockConnectionFactory` - but it's quite a bit more work than I thought it would be. Maybe over time...